src/go/overlord/statemanager.go

```diff
@@ -0,0 +1,168 @@
+//  Copyright 2018 Google LLC
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+
+//        https://www.apache.org/licenses/LICENSE-2.0
+
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//	limitations under the License.
+
+package overlord
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	mpb "github.com/google/minions/proto/minions"
+	pb "github.com/google/minions/proto/overlord"
+	"github.com/patrickmn/go-cache"
+)
+
+// StateManager handles the state of an Overlord through multiple
+// scans.
+type StateManager interface {
+	// AddFiles atomically sets the state of a minion during a scan.
+	AddFiles(scanID string, files []*pb.File) error
+	// AddInterest adds a new interest for a given minion to the state of the scan.
+	AddInterest(scanID string, interest *mpb.Interest, minion string) error
+	// CreateScan initializes the state for a scan.
+	CreateScan(scanID string) error
+	// GetFiles returns all the files known for a given ScanID
+	GetFiles(scanID string) ([]*pb.File, error)
+	// GetInterests returns all the interests known for a given ScanID, mapped to minions
+	GetInterests(scanID string) ([]*mappedInterest, error)
+	// RemoveFile atomically removes a given file from the state.
+	RemoveFile(scanID string, file *pb.File) error
+	// ScanExists returns true if any state at all is known about the scan.
+	ScanExists(scanID string) bool
+}
+
+// LocalStateManager handles state through a local time-expiring cache.
+type LocalStateManager struct {
```

Comment:
You could move this to a separate package (say, 'state'), and name this Local. Then calling this would be state.NewLocal(), and in future also state.NewFoobar where Foobar is some form of a distributed state manager.

Reply:
I think this is a good idea. One drawback though is that mappedInterest is now caught between the 2 packages (it's defined in the overlord, I was using it in state). Any suggestions on how to break this?

Reply:
Ok, I've moved everything into the state package :-) Let me know what you think